### PR TITLE
Fix Integrator and Neighbor List detaching

### PR DESCRIPTION
**Pull request by b-butler**
## Description

Fixes `hoomd.md._DynamicIntegrator` and `hoomd.md.nlist.NList` detaching. Currently MD integrator objects do not detach any of the forces, methods, or constraints or the rigid parameter. The neighbor list is not detached by the composing potentials either. This is fixed as well through using the `_DependencyRelation` interface. This interface is cleaned up and fixed in this PR as well.

This PR cherry-picks a commit from #1099, so either a rebase and force push here or there will be needed after the other is merged.

This is also dependent on #1053 like #1099.

## Motivation and context

This fixes multiple existing bugs with detaching properly which is done when operations or other objects are removed from a simulation that is attached.

## How has this been tested?

Currently, neighbor list detaching is tested, but no tests for the md integrator currently exist. I can add the test for detaching, but I don't think all the tests should go in this PR.

## Change log

```
Fixes: Removing an integrator from a simulation after running.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/md/nlist.py [hunk 1]
```diff
@@ -82,12 +82,6 @@ class NList(_HOOMDBaseObject):
             :math:`[\mathrm{length}]`.
     """
 
-    _remove_for_pickling = _HOOMDBaseObject._remove_for_pickling + (
-        '_cpp_cell',)
-    _skip_for_equality = _HOOMDBaseObject._skip_for_equality | {
-        '_cpp_cell',
-    }
-
```
[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/md/nlist.py:90]: We don't ever set `_cpp_cell` as far as I can see.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! We need to merge master into this branch and fix the failing tests before merging.

File: hoomd/md/methods/methods.py [hunk 2]
```diff
@@ -23,15 +23,7 @@ class Method(_HOOMDBaseObject):
     Note:
         Users should use the subclasses and not instantiate `Method` directly.
     """
-
-    def _detach(self):
```
Comment (joaander) [inline hoomd/md/methods/methods.py:27]: Is this now handled in the base class?

**Review by joaander — APPROVED** [review]
Thanks! I have one question for clarification.

**Comment by b-butler** [comment]
@joaander, I am going to rerequest review because some major changes were necessary to get the desired behavior implemented. Some may seem unrelated, but they aren't due to requirements for avoiding copying the neighbor list unless necessary. The changes are generally cleaner regardless. I added new tests to ensure the behavior is as expected.

File: hoomd/md/methods/methods.py (hunk 2 above)
Comment (b-butler) [reply hoomd/md/methods/methods.py:27]: Yes, this keeps popping back up some how, I deleted it in another branch as well.

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**